finetune/base.py

```diff
@@ -194,56 +101,91 @@ def _target_model(self, *, featurizer_state, targets, n_outputs, train=False, re
         raise NotImplementedError
 
-    @abstractmethod
-    def _target_encoder(self):
-        # Overridden by subclass to produce the right target encoding for a given target model.
-        raise NotImplementedError
-
-    def _eval(self, *tensors, feed_dict):
-        """
-        Evaluate the value of each of the provided tensors.
-        Returns a `dict` that maps from tensor to result value.  
-        If any result value is None, that result is excluded from the results `dict`.
-        """
-        tensors = [
-            tensor if tensor is not None else self.noop
-            for tensor in tensors
-        ]
-        values = self.sess.run(tensors, feed_dict=feed_dict)
-        return {
-            tensor: value
-            for tensor, value in zip(tensors, values)
-            if value is not None
-        }
-
     def finetune(self, Xs, Y=None, batch_size=None):
-        if Y is not None and len(Xs) != len(Y):
+        if not callable(Xs) and Y is not None and len(Xs) != len(Y):
             raise FinetuneError(
                 "Mismatch between number of examples ({}) and number of targets ({}) provided.".format(
                     len(Xs),
                     len(Y)
                 )
             )
-        arr_encoded = self._text_to_ids(Xs)
-        return self._training_loop(
-            arr_encoded,
-            Y=Y,
-            batch_size=batch_size,
-        )
+        batch_size = batch_size or self.config.batch_size
+        val_size, val_interval = self.validation_settings(
+            n_examples=len(Xs) if not callable(Xs) else self.config.dataset_size,
+            batch_size=batch_size or self.config.batch_size)
+        if val_size == 0:
+            val_interval = sys.maxsize
+        val_input_fn, train_input_fn = self.input_pipeline._get_train_input_fns(Xs, Y, batch_size=batch_size,
+                                                                                val_size=val_size)
+
+        train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn)
+        eval_spec = tf.estimator.EvalSpec(input_fn=val_input_fn)
+        estimator = self.get_estimator(val_interval=val_interval)
+        tf.estimator.train_and_evaluate(estimator, train_spec=train_spec,
+                                        eval_spec=eval_spec)
+
+
+    def get_estimator(self, val_interval=None, force_build_lm=False):
+        if self.estimator_ is None or self.input_pipeline.rebuild or force_build_lm:
+            hot_start = None
+            if self.input_pipeline.rebuild:
+                try:
+                    hot_start = tf.estimator.WarmStartSettings(
+                        ckpt_to_initialize_from=tf.train.latest_checkpoint(self.estimator_dir),
+                        vars_to_warm_start=[name for name, _ in tf.train.list_variables(self.estimator_dir) if
+                            not name.startswith("OptimizeLoss")])
+                    os.remove(os.path.join(self.estimator_dir, "checkpoint"))
```

Comment:
Wait does this remove the cptk file that the model was just loaded from?  Does this potentially just nuke your ckpt file and leave you high and dry if an error is raised before a new ckpt file is written?